protoc-gen-flowtypes/genflowtypes/genflowtypes.go

```diff
@@ -6,73 +6,97 @@ import (
 	"strings"
 	"text/template"
 
+	"github.com/golang/protobuf/proto"
 	pbdescriptor "github.com/golang/protobuf/protoc-gen-go/descriptor"
 	"github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway/descriptor"
+	"github.com/mwitkow/go-proto-validators"
 )
 
-type FlowType interface {
+// Flower is a flow language type
+type Flower interface {
 	FlowType() string
+	IsRequired() bool
 }
-type NamedFlowType interface {
-	FlowType
-	FlowTypeName() string
+
+// NamedFlower is a Flower with a name
+type NamedFlower interface {
+	Flower
+	Name() string
 }
 
-type simpleFlowType string
+func newSimpleType(typeString string, required bool) *simpleType {
+	return &simpleType{typeString, required}
+}
 
-func (s simpleFlowType) FlowType() string { return string(s) }
+type simpleType struct {
+	typeString string
+	required   bool
+}
 
-type messageFlowType string
+func (s simpleType) FlowType() string { return s.typeString }
+func (s simpleType) IsRequired() bool { return s.required }
 
-func (s messageFlowType) FlowType() string { return string(s) }
+type messageType struct {
+	*simpleType
+}
 
-type repeatedFlowType struct {
-	t FlowType
+func newMessageFlowType(typeName string, required bool) *messageType {
+	s := newSimpleType(typeName, required)
+	return &messageType{s}
 }
 
-func (r repeatedFlowType) FlowType() string { return fmt.Sprintf("Array<%s>", r.t.FlowType()) }
+type repeatedType struct {
+	Flower
+	required bool
+}
 
-type namedFlowType struct {
-	Name string
-	Type FlowType
+func newRepeatedFlowType(underlying Flower, required bool) *repeatedType {
+	return &repeatedType{Flower: underlying, required: required}
 }
 
-func (t *namedFlowType) FlowType() string {
-	return t.Type.FlowType()
-	return fmt.Sprintf("%s = %s", t.Name, t.Type.FlowType())
+func (r repeatedType) FlowType() string { return fmt.Sprintf("Array<%s>", r.Flower.FlowType()) }
+
+type namedType struct {
+	Flower
+	name     string
+	required bool // TODO(tmc): move to interface
 }
-func (t *namedFlowType) FlowTypeName() string {
-	return t.Name
+
+func (t *namedType) Name() string {
+	return t.name
 }
 
 type objectFlowType struct {
-	Fields  []NamedFlowType
-	Options Options
+	Fields   []NamedFlower
```

Comment:
yes I believe so, I'm not sure about the relative counts of those in the wild, why?